File: clients/android/core/sync/src/main/kotlin/com/clipshare/core/sync/EpochDistributionPlan.kt
package com.clipshare.core.sync

import com.clipshare.core.vault.DeviceId

data class PairedDeviceState(val deviceId: DeviceId, val revoked: Boolean)

class EpochDistributionPlan private constructor(
    val previousEpoch: Long,
    val newEpoch: Long,
    val recipients: Set<DeviceId>,
) {
    fun validateGeneratedWrappers(wrapperRecipients: Set<DeviceId>) {
        require(wrapperRecipients == recipients) {
            "A new epoch must be wrapped exactly for the non-revoked device set."
        }
    }

    companion object {
        fun create(previousEpoch: Long, devices: Collection<PairedDeviceState>): EpochDistributionPlan {
            require(previousEpoch > 0 && previousEpoch < Long.MAX_VALUE) { "The previous epoch must be rotatable." }
            require(devices.map(PairedDeviceState::deviceId).toSet().size == devices.size) {
                "Paired devices must be unique."
            }
            return EpochDistributionPlan(
                previousEpoch,
                previousEpoch + 1,
                devices.filterNot(PairedDeviceState::revoked).mapTo(mutableSetOf(), PairedDeviceState::deviceId),
            )
        }
    }
}

File: clients/windows/src/ClipShare.Windows.Vault/EpochDistributionPlan.cs
namespace ClipShare.Windows.Vault;

public sealed record PairedDeviceState(DeviceId DeviceId, bool Revoked);

public sealed record EpochDistributionPlan
{
    private EpochDistributionPlan(long previousEpoch, long newEpoch, IReadOnlySet<DeviceId> recipients)
    {
        PreviousEpoch = previousEpoch;
        NewEpoch = newEpoch;
        Recipients = recipients;
    }

    public long PreviousEpoch { get; }

    public long NewEpoch { get; }

    public IReadOnlySet<DeviceId> Recipients { get; }

    public static EpochDistributionPlan Create(long previousEpoch, IReadOnlyCollection<PairedDeviceState> devices)
    {
        ArgumentNullException.ThrowIfNull(devices);
        ArgumentOutOfRangeException.ThrowIfNegativeOrZero(previousEpoch);
        if (previousEpoch == long.MaxValue)
        {
            throw new InvalidOperationException("The epoch number is exhausted.");
        }

        if (devices.Select(device => device.DeviceId).Distinct().Count() != devices.Count)
        {
            throw new ArgumentException("Paired devices must be unique.", nameof(devices));
        }

        return new EpochDistributionPlan(
            previousEpoch,
            previousEpoch + 1,
            devices.Where(device => !device.Revoked).Select(device => device.DeviceId).ToHashSet());
    }

    public void ValidateGeneratedWrappers(IReadOnlySet<DeviceId> wrapperRecipients)
    {
        ArgumentNullException.ThrowIfNull(wrapperRecipients);
        if (!Recipients.SetEquals(wrapperRecipients))
        {
            throw new InvalidOperationException("A new epoch must be wrapped exactly for the non-revoked device set.");
        }
    }
}
